fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/IfFunctionTest.java

```diff
@@ -0,0 +1,384 @@
+
+/**
+ * @author John Grimes
+ */
+@SpringBootTest
+@Tag("UnitTest")
+class IfFunctionTest {
+
+  @Autowired
+  private SparkSession spark;
+
+  @Autowired
+  private FhirContext fhirContext;
+
+  private ParserContext parserContext;
+  private ResourceReader resourceReader;
+
+  @BeforeEach
+  void setUp() {
+    parserContext = new ParserContextBuilder(spark, fhirContext).build();
+    resourceReader = mock(ResourceReader.class);
+  }
+
+  @Test
+  void returnsCorrectResultsForTwoLiterals() {
+    final Dataset<Row> inputContextDataset = new ResourceDatasetBuilder(spark)
+        .withIdColumn()
+        .withRow("observation-1")
+        .withRow("observation-2")
+        .withRow("observation-3")
+        .withRow("observation-4")
+        .withRow("observation-5")
+        .build();
+    when(resourceReader.read(ResourceType.OBSERVATION)).thenReturn(inputContextDataset);
+    final ResourcePath inputContext = new ResourcePathBuilder(spark)
+        .expression("Observation")
+        .resourceType(ResourceType.OBSERVATION)
+        .resourceReader(resourceReader)
+        .singular(true)
+        .build();
+    final Dataset<Row> inputDataset = new DatasetBuilder(spark)
+        .withIdColumn()
+        .withEidColumn()
+        .withColumn(DataTypes.BooleanType)
+        .withRow("observation-1", makeEid(0), true)
+        .withRow("observation-2", makeEid(0), false)
+        .withRow("observation-3", makeEid(0), null)
+        .withRow("observation-4", makeEid(0), true)
+        .withRow("observation-4", makeEid(1), false)
+        .withRow("observation-5", makeEid(0), null)
+        .withRow("observation-5", makeEid(1), null)
+        .build();
+    final ElementPath inputPath = new ElementPathBuilder(spark)
+        .fhirType(FHIRDefinedType.BOOLEAN)
+        .dataset(inputDataset)
+        .idAndEidAndValueColumns()
+        .expression("valueBoolean")
+        .singular(false)
+        .build();
+    final NonLiteralPath condition = inputPath.toThisPath();
+    final StringLiteralPath ifTrue = StringLiteralPath.fromString("foo", inputContext);
+    final StringLiteralPath otherwise = StringLiteralPath.fromString("bar", inputContext);
+
+    final NamedFunctionInput ifInput = new NamedFunctionInput(parserContext, inputPath,
+        Arrays.asList(condition, ifTrue, otherwise));
+    final FhirPath result = NamedFunction.getInstance("iif").invoke(ifInput);
+
+    final Dataset<Row> expectedDataset = new DatasetBuilder(spark)
+        .withIdColumn()
+        .withEidColumn()
+        .withColumn(DataTypes.StringType)
+        .withRow("observation-1", makeEid(0), "foo")
+        .withRow("observation-2", makeEid(0), "bar")
+        .withRow("observation-3", makeEid(0), "bar")
+        .withRow("observation-4", makeEid(0), "foo")
+        .withRow("observation-4", makeEid(1), "bar")
+        .withRow("observation-5", makeEid(0), "bar")
+        .withRow("observation-5", makeEid(1), "bar")
+        .build();
```

Comment:
I can see that we treat null values in the condition as 'false' but is that correct? 
Should we not rather preserve nulls?

Reply:
I think that is the case for most FHIRPath functions, but the spec for this one says:

> If criterion is false or an empty collection, the function returns otherwise-result, unless the optional otherwise-result is not given, in which case the function returns an empty collection.

Reply:
Note that we don't do the "optional" part currrently - I'm not really sure what the value of it is.

Doesn't leaving out the third argument just turn this into `where`?